From c9e7d7025deb7094923fb86031cbe499a5a85e24 Mon Sep 17 00:00:00 2001
From: Pedro Luis Rodríguez González <pl.rguez@gmail.com>
Date: 24 ago. 2020 16:16:29

Patch #012 for OpenDingux and RetroFW
Separated GCW0 General options.
Show joystick in use at status info.
New option for show throughput in % or fps.
Spacing.

diff --git a/Platform/readme.txt b/Platform/readme.txt
index 5c84440..2f113d1 100644
--- a/Platform/readme.txt
+++ b/Platform/readme.txt
@@ -1,13 +1,14 @@
 Using the Fuse emulator
 =======================
 
-This port of Fuse to OpenDingux and RetroFW is simply a compilation of Fuse 1.5.7 for this platforms with some additions to the SDL UI and some new options.
-
-Nothing is touched for the emulation engine or the miriad of Fuse options.
+This port of Fuse to OpenDingux and RetroFW is a compilation of Fuse 1.5.7 for Opendingux/RetroFW
+with some additions to the SDL UI and some new options.
 
 So the thanks for this port are for Phillip Kendall and all the Fuse team.
+And also to the GCW0/OpenDingux and RetroFW community for all documentation, development threads, and
+all opensource ports.
 
-At the end of this doc is the man page of Fuse 1.5.7 for reference.
+At the end of this doc is the man page of Fuse 1.5.7 for a complete reference.
 
 ------------------------
 ### Opening the Menu ###
@@ -23,12 +24,17 @@
 
 By default the emulator don't save settings changes. You can save settings at any time from menu: `Menu -> Options -> Save`.
 
-You can activate "Auto-save" to save settings when you exit the emulator. To activate it access to `Menu -> Options -> General` then navigate to the option **Auto-save settings**, check it with the `X` Button and finally accept the form with the `A` Button. Yo should exit from the emulator o use the Save option to save it.
+You also can activate "Auto-save" to save settings when you exit the emulator.
 
-The options are saved into 'fuse.cfg' file in the fuse config path. If some option changes prevent to open the emulator delete this file and try to start again.
-The config path is '$HOME/.fuse' directory. If it does not exist the emulator will create it at start.
+|To activate auto-save access to `Menu -> Options -> General` and check 'Auto-save settings'.
+|Check the option with the `X` button and accept options with the `A` button.
+|You should exit from the emulator or use the 'Save' to save this changes.
 
-Note: $HOME is located in '/media/data/local/home' for OpenDingux and in '/home/retrofw' for RetroFW.
+In this port the options are saved into 'fuse.cfg' file in the fuse config path.
+If some option changes prevent to open the emulator delete this file and try to start again.
+
+The config path is in '$HOME/.fuse' directory. If it does not exist the emulator will create it at start.
+|$HOME is located in '/media/data/local/home' for OpenDingux and in '/home/retrofw' for RetroFW.
 
 ------------------------
 ### Virtual Keyboard ###
@@ -36,34 +42,26 @@
 
 Press the `Start` button to open a Virtual Keyboard when you are at emulator.
 
-The keys in the virtual keyboard correspond to ZX Spectrum keys and it send to emulator as if you are using a real ZX Spectrum keyboard.
+The keys in keyboard correspond to ZX Spectrum 48k keys.
+When you press them are send to emulator as if you are using a real ZX Spectrum keyboard.
 
-    `Cursor`
-    Move to select keys.
+  `Cursor`    Move to select keys.
+  `A`         Press selected key.
 
-    `A`
-    Press selected key.
+  `B`         Lock key. It's marked in blue.
+              The locked keys are used when press the `A` button over a Key and then are unlocked.
+              For use in combinations of Caps Shift or Symbol Shift with other keys.
+              For example: In 48k mode you can Lock key "Ss" and then press 'Cs' key to change keyboard to "Extended" mode.
 
-    `B`
-    Lock key. It's marked in blue.
-    The locked keys are used when press the `A` button over a Key and then are unlocked.
-    For use in combinations of Caps Shift or Symbol Shift with other keys.
-    For example: In 48k mode you can Lock key "Ss" and then press 'Cs' key to change keyboard to "Extended" mode.
+  `X`         Sticky key. It's marked in Red.
+              The key is continously pressed.
+              For example: In 48k mode you can press key 'J' then Sticky key 'Ss' and then press twice key 'P' to obtain 'LOAD ""'.
 
-    `X`
-    Sticky key. It's marked in Red.
-    The key is continously pressed.
-    For example: In 48k mode you can press key 'J' then Sticky key 'Ss' and then press twice key 'P' to obtain 'LOAD ""'.
+  `Y`         Clean all Lock and Sticky keys.
+  `Start`     Close Virtual Keyboard
 
-    `Y`
-    Clean all Lock and Sticky keys.
-
-    `Start`
-    Close Virtual Keyboard
-
-    `L1`, `R1`
-    Change the Virtual Keyboard position. 
-    There are 4 positions, one for earch corner of screen.
+  `L1`, `R1`  Change the Virtual Keyboard position.
+              There are 4 positions, one for earch corner of screen.
 
 -------------------------------------
 ### Buttons at different contexts ###
@@ -76,49 +74,39 @@
 
 The buttons function detailed are when no mapping is assigned.
 
-    `Cursor`
-    Shifted cursor ZX Spectrum keys (5, 6, 7 and 8 keys)
-    This is modified by the setting "Use shift with arrow keys"
+  `Cursor`    Shifted cursor ZX Spectrum keys (5, 6, 7 and 8 keys)
+              This is modified by the setting "Use shift with arrow keys"
 
-    `A`, `B`
-    Symbol Shift
+  `A`, `B`    Symbol Shift
+  `Y`         CAPS Shift
+  `X`         Space
 
-    `Y`
-    CAPS Shift
+  `Select`, `Power`
+              Open Menu
 
-    `X`
-    Space
+  `Start`     Open/Close the virtual keyboard
 
-    `Select`, `Power`
-    Open general menu
+  `L1`        If hotkey combos disabled: Nothing
+              If hotkey combos enabled: Start a hotkey combo
 
-    `Start`
-    Open the virtual keyboard
+  `R1`        If hotkey combos disabled: Caps Shith + 0 (DELETE)
+              If hotkey combos enabled: Start a hotkey combo
 
-    `L1`
-    If hotkey combos disabled: Nothing
-    If hotkey combos enabled: Start a hotkey combo
+  `Left Stick`
+              Nothing.
+              On some systems you can map it as cursor.
 
-    `R1`
-    If hotkey combos disabled: Caps Shith + 0 (DELETE)
-    If hotkey combos enabled: Start a hotkey combo
+  `Right Stick`
+              Nothing.
+              On some systems you can enable an emulated mouse that uses the left
+              stick to move the mouse. In Fuse you can use it as Kempston mouse.
+              |In RG350 you can enable it with `Power`+`B` hotkey.
 
-    `Left Stick`
-    Nothing.
-    On some systems you can map it as cursor.
+  `L2`        Nothing.
+              Left click with mouse emulation enabled.
 
-    `Right Stick`
-    Nothing. 
-    On some systems you can enable an emulated mouse that uses the left stick to move the mouse. In Fuse you can use it as Kempston mouse.
-    |In RG350 you can enable it with `Power`+`B` hotkey.
-
-    `L2`
-    Nothing.
-    Left click with mouse emulation enabled.
-
-    `R2`
-    Nothing.
-    Right click with mouse emulation enabled.
+  `R2`        Nothing.
+              Right click with mouse emulation enabled.
 
 
 Buttons in Menu:
@@ -126,46 +114,33 @@
 
 These are no dependent of button mapping.
 
-    `Cursor`, `Left Stick`
-    Move
+  `Cursor`, `Left Stick`
+             Move
 
-    `A`
-    Select the option or accept the options at form
+  `A`        Select the option or accept the options at form
+  `B`        Cancel and go back to previous menu or to the emulator if there is no previous menu
 
-    `B`
-    Cancel and go back to previous menu or to the emulator if there is no previous menu
+  `X`        At different contexts:
+               - Mark/Unmark for check options: For example at General options -> Auto-save settings
+               - Open list of options for list options: For example at Sound options -> AY stereo separation
+               - It open 'enter name' dialog for file save dialogs.
 
-    `X`
-    At different contexts:
-    - Mark/Unmark for check options: For example at General options -> Auto-save settings
-    - Open list of options for list options: For example at Sound options -> AY stereo separation
-    - It open 'enter name' dialog for file save dialogs.
+  `L1`       Go to the first menu option
+  `R1`       Go to the last menu option
+  `L2`       In the file selector go to the first entry in the current directory
+  `R2`       In the file selector go to the last entry in the current directory
+  `Power`    Cancel and exit completely from any level of menu to the emulator
 
-    `L1`
-    Go to the first menu option
+  `Y`, `Select`, `Start`, `L2`, `R3`, `Right Stick`
+             Nothing
 
-    `R1`
-    Go to the last menu option
+#### Maintain options ####
+--------------------------
+Some menu options open a form with options to check/uncheck or choose option from a list.
 
-    `L2`
-    In the file selector go to the first entry in the current directory
-
-    `R2`
-    In the file selector go to the last entry in the current directory
-
-    `Power`
-    Cancel and exit completely from any level of menu to the emulator
-
-    `Y`, `Select`, `Start`, `L2`, `R3`, `Right Stick`
-    Nothing
-
-#### Forms ####
----------------
-Some menu options open a form with options to mark or choose option from a list.
-
-The forms must be confirmed to persist the selected options
-  To confirm them you must use the `A` button.
-  The `B` or `Power` buttons cancel them and the changes are lost.
+  - To check/uncheck options use `X` button.
+  - To confirm changes in form use the `A` button.
+  - To cancel changes in form use the `B` or `Power` buttons.
 
 Some examples of forms are General, Media and Sound.
 
@@ -176,23 +151,14 @@
 
 Examples where the virtual keyboard it's appearing are: 'Enter name' for save dialogs, Debugger, Poke Finder, Poke Memory.
 
-    `Cursor`
-    Move
+  `Cursor`   Move to select keys
+  `A`        Press the selected key
+  `X`        Change the keyboard mode between upper and lower case and some additional characters
+  `Y`        Delete previous character
+  `B`        Cancel and exit to previous option in menu
 
-    `A`
-    Press the selected key
-
-    `X`
-    Change the keyboard mode between upper and lower case and some additional characters
-
-    `Y`
-    Delete previous character
-
-    `B`
-    Cancel and exit to previous option in menu
-
-    Confirm input
-    To confirm press 'En' key at virtual Keyboard
+  Confirm input
+             To confirm the input press 'En' key at virtual Keyboard
 
 |For the 'Enter the name' in save dialogs the last loaded filename without the extension is proposed.
 
@@ -202,36 +168,32 @@
 
 You can map handheld buttons to Joysticks or Spectrum keys in `Menu -> Options -> Joysticks`
 
----- GCW0 Joystick 1... ----
-----------------------------
+==== GCW0 Joystick 1 ====
 This option allows to emulate a ZX Spectrum joystick with the handheld buttons.
 It also allows you to assign keyboard keys to buttons
 
-- **Type**:
-  Choose the Joystick to emulate or None. Default is None.
+  - Type: Choose the Joystick to emulate or None. Default is None.
+  - Button mapping: Map the handheld buttons to Joystick fire, ZX Spectrum keys or to Nothing.
 
   Some type of Joysticks may require other emulator options enabled.
-  |For example for Kempston joystick emulation you need to enable 'Kempston joystick' in `Menu --> Options --> Peripherals --> General`.
+  |For example for Kempston joystick emulation you need to enable 'Kempston joystick'
+  |in `Menu --> Options --> Peripherals --> General`.
+  By default all buttons are mapped to Joystick fire.
 
-- **Button mapping**:
-  Map the handheld buttons to Joystick fire, ZX Spectrum keys or to Nothing. By default all buttons are mapped to Joystick fire.
-
----- GCW0 Keyboard... ----
---------------------------
+==== GCW0 Keyboard ====
 This option allows to map handheld buttons to ZX Spectrum keys. The cursors also can be mapped.
 
-- **Type**:
-  Choose Activated or None. Default is None.
+  - Type: Choose Activated or None. Default is None.
+  - Button mapping: Map buttons to ZX Spectrum keys or Nothing. By default all buttons are mapped to Nothing.
 
-- **Button mapping**:
-  Map buttons to ZX Spectrum keys or Nothing. By default all buttons are mapped to Nothing.
+--------------------------
 
 When both mappings, "GCW0 Joystick 1" and "GCW0 Keyboard", are enabled at the same time only "GCW0 Joystick 1" will work.
 
-When you choose any 'Type' other than "None" in "GCW0 Joystick 1" or "GCW0 Keyboard", some buttons may lose their original functionality if they are mapped to joystick fire or to keyboard keys.
-`Select`, `Start`, `L1` and `R1` buttons have functionanility that may be lost.
-
-If you have mapped the `Select` button, you can still access the menu with the` Power` button.
+When you choose any 'Type' other than "None" in "GCW0 Joystick 1" or "GCW0 Keyboard", some buttons may lose their original functionality if they are
+mapped to joystick fire or to keyboard keys.
+  - `Select`, `Start`, `L1` and `R1` buttons have functionanility that may be lost.
+  - If you have mapped the `Select` button, you can still access the menu with the` Power` button.
 
 ----------------------------------------------
 ### Control mapping per game configuration ###
@@ -239,19 +201,28 @@
 
 To have control gaming configuration per game you must activate it in the menu 'Options -> Joysticks -> Control mapping'.
 
-See PITFALLs at end of this section.
+Files for save control mapping:
+  - The configuration files per game are saved in 'mappings' directory in fuse config path ($HOME/.fuse/mappings).
+  - They have extension '.fcm' (fcm is for Fuse Control Mapping).
+  - The options saved are the 'Types' and buttons mappings for GCW0 Joystick 1, Joystick 2, Keyboard and GCW0 Keyboard.
+  - The format of the file is the same used for the general settings: XML for OpenDingux or plain text for RetroFW.
+  - By default for the control mapping filenames will be tried to detect some patterns on name to cut them off:
+        - All the denominations between '()' or '[]'
+        - 'Tape', 'Disk', 'Side' for 'ABCD' or '1234', 'Part 1234 of 1234'
+        - The '128k', '48k' out of '()' '[]'
+        - The 'Small, Medium, Large' whatever 'Case'
+        (The search of patterns are case insensitive)
 
+The supported media for auto-load control mapping files are tapes, microdrives, snapshots, disks (not IDE), rom cartridges and Timex cartridges.
+
+See PITFALLs at the end of this section.
+
+Options:
   - Control mapping per game:
     -------------------------
     Default disabled.
     It enable the save and load of control mappings per game for supported media: : tapes, snapshots, disks (not IDE), rom cartridges, Timex cartridges.
 
-    The configuration files per game are saved in 'mappings' directory in fuse config path ($HOME/.fuse/mappings).
-    The name of the files will be the same of the file loaded replacing its extenstion by '.fcm' (fcm is for Fuse Control Mapping).
-    The options saved are the 'Types' and buttons mappings for GCW0 Joystick 1, Joystick 2, Keyboard and GCW0 Keyboard.
-
-    The format of the file is the same used for the general settings: XML for OpenDingux or plain text for RetroFW.
-
     If you disable it:
       - If defaults are detached they are saved and restored as current defaults.
       - If changes were made to controls those changes are saved to correponding control mapping file.
@@ -264,7 +235,7 @@
     ----------
     Default enabled. Only take effect if 'Control mapping per game' is enabled.
 
-    It enable the autoload of control mapping files associated to the file loaded.
+    If enabled the autoload of control mapping files associated to the file loaded.
     This apply for supported media when are loaded from 'File -> Open' or inserted from the 'Media' menu.
 
   - Auto-save:
@@ -281,18 +252,19 @@
     ----------------------
     Default disabled. Only take effect if 'Control mapping per game' is enabled.
 
-    This allow to maintain default control mapping different of particular of each game.
-
-    If this option it's not enabled then you can maintain separated default controls.
+    If this option is enabled you can't maintain separated default controls.
 
     The default controls are used:
       - At start of Fuse, if no media is autoload or no control mapping autoload is enable
       - When you insert a media with not yet control mapping configuration
       - When you clear or eject a media
 
-    If this option is enabled the changes to controls are the defaults. 
-      - Loading any control mapping file, automatically or manually change it
-      - If you load any media with no control mapping yet then this defaults will be used.
+  - No cut/transform filenames:
+    ---------------------------
+    Default disabled. Only take effect if 'Control mapping per game' is enabled.
+
+    With this option enabled the filename for mapping control will be the same of the file loaded replacing the extension by '.fmc'.
+    With this option disabled the control mapping filename follow the rules indicated above.
 
 When you exit from Fuse and have General Auto-save setting enabled then default control mapping will be saved at general file.
 
@@ -346,14 +318,8 @@
 
 PITFALLS:
   Variants of the same program:
-    Control mapping filenames are based on the filename loaded.
-    For games with two or more files (tapes, disk sides) or variations of the same program (48k, 128k, different publishers or formats) the emulator try to detect some patterns on name to cut them off:
-      - All the denominations between '()' or '[]'
-      - 'Tape', 'Disk', 'Side' for 'ABCD' or '1234', 'Part 1234 of 1234'
-      - The '128k', '48k' out of '()' '[]'
-      - The 'Small, Medium, Large' whatever 'Case'
-    The searchs are case insensitive.
-    But if the names used are no consistent then a different name for the same program may be determined.
+    If 'No cut/transform filenames' is disabled (default) the emulator try to clean the name of filename to make it equal name for variations.
+    But if the filenames used are no consistent, including letter case, then a different name for the same program may be determined.
 
   When insert various media the last inserted will be the selected control mapping.
     - If you unload it then it will be cleared and you could start a game from other media inserted with no file control mapping.
@@ -375,38 +341,24 @@
 
 On some handhelds it is possible to connect an external Joystick using USB OTG. The mapping of 'Joystick 2' will be applied to it.
 
-|On my RG350 I've tested this with an 8Bitdo FC30 Pro controller.
-
 For controllers the number of each button in mapping depends on the controller layout.
-For example on my 8Bitdo FC30 these are the mapping layout: A=1, B=2, X=4, Y=5, L1=7, R1=8, L2=9, R2=10, Select=11, Start=12, L3=14, R3=15.
-From SDL Game controller DB (https://github.com/gabomdq/SDL_GameControllerDB/blob/master/gamecontrollerdb.txt)
-	"03000000c82d00001038000000000000,8BitDo FC30 Pro,
-	a:b0,
-	b:b1,
-	back:b10,
-	dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,
-	leftshoulder:b6,
-	leftstick:b13,
-	lefttrigger:b8,
-	leftx:a0,lefty:a1,
-	rightshoulder:b7,
-	rightstick:b14,
-	righttrigger:b9,
-	rightx:a3,
-	righty:a4,
-	start:b11,
-	x:b3,
-	y:b4,
-	hint:SDL_GAMECONTROLLER_USE_BUTTON_LABELS:=1,"
 
 =>Fuse count buttons from 1 not from 0, so add 1 to translate your layout into Fuse's config.
 
+|On my RG350 I've tested this with an 8Bitdo FC30 Pro controller.
+|The mapping layout: A=1, B=2, X=4, Y=5, L1=7, R1=8, L2=9, R2=10, Select=11, Start=12, L3=14, R3=15.
+|From SDL Game controller DB (https://github.com/gabomdq/SDL_GameControllerDB/blob/master/gamecontrollerdb.txt)
+    "03000000c82d00001038000000000000,8BitDo FC30 Pro,
+        a:b0,b:b1,back:b10,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,
+        leftshoulder:b6,leftstick:b13,lefttrigger:b8,leftx:a0,lefty:a1,
+	rightshoulder:b7,rightstick:b14,righttrigger:b9,rightx:a3,righty:a4,start:b11,x:b3,y:b4,
+	hint:SDL_GAMECONTROLLER_USE_BUTTON_LABELS:=1,"
+
 -------------------------
 ### External keyboard ###
 -------------------------
 
 I don't have tested a Keyboard on USB OTG but in theory it should "work".
-
 But keep in mind that current adaptations made for handhelds buttons have probably made it unusable.
 
 The handheld buttons are keystrokes that have conflict with some keys used for SDL menus.
@@ -416,9 +368,7 @@
 ----------------------
 
 In some handhelds you can emulate mouse with the Right stick and `L2`, `R2` buttons.
-
 This make possible emulating the Kempston mouse.
-
 You must enable 'Kempston mouse' in the peripherals: `Menu -> Options -> Peripherals -> General`, and enable the mouse emulation in the handheld.
 
 |For example, in the RG350 you can activate the mouse with hotkey `Power` + `B`.
@@ -427,7 +377,7 @@
 ### Triple Buffer ###
 ---------------------
 
-There is an implementation to use the triple buffer. It can be enabled at General options or whith combo hotkeys 'L1' + 'R1' + 'B'.
+There is an implementation to use the triple buffer. It can be enabled at 'General GCW0' options or whith combo hotkeys 'L1' + 'R1' + 'B'.
 
 If tripple buffer is activated then an [B] will be at status line.
 
@@ -442,7 +392,7 @@
 ### Hotkey combinations ###
 ---------------------------
 
-For use hotkey combos it they must be enabled in Menu -> Options -> General.
+For use hotkey combos it they must be enabled in 'Menu -> Options -> General GCW0'.
 
 If `L1` and `R1` buttons are not mapped to Joystick or Keyboard it will be used to start the hotkey combinations.
 
@@ -450,28 +400,23 @@
 
     L1 + R1 + B      Toggle triple buffer
     L1 + R1 + X      Joystick
-
     L1 + Select + Y  Tape play (F8)
-
     L1 + A           Tape open (F7)
     L1 + B           Save file (F2)
     L1 + X           Open file (F3)
     L1 + Y           Media menu
-
     R1 + A           General options (F4)
     R1 + B           Reset machine (F5)
     R1 + X           Exit fuse (F10)
     R1 + Y           Machine select (F9)
 
-(Yes, I know, some combinations on some devices are impracticable)
-
 ---------------------
 ### Media options ###
 ---------------------
 
 If you have problems loading some programs keep in mind the compatibility issues of software with the model emulated.
 
-Also some options in Media options can influence to some loaders as "Accelerate loaders". Try enabling/disabling this options.
+Also some options in 'Media' options can influence to some loaders as 'Accelerate loaders'. Try enabling/disabling different otions.
 
 ------------
 ### ROMs ###
@@ -479,7 +424,7 @@
 
 Fuse provides roms for some systems and peripherals it emulate but not for all models or peripherals that it support.
 
-You can assign not distributed roms or change the default roms from the emulator itself:
+You can assign new roms or change the defaults from the emulator itself:
 
 - Open menu with `Select` or `Power` button. Then navigate from Menu --> Options --> Select Roms --> Machine ROMs or Peripheral ROMs
 - With `cursor` select the rom to change and press `X` button
@@ -487,14 +432,14 @@
 - Repeat until all the needed ROMs are assigned.
 - Finally do accept all pressing the `A` button.
 
-You can reset a rom to Fuse's default rom with the `Y` button.
+You can reset a rom to default with the `Y` button.
 
 NOTE:
   The emulator detect your rom as custom if they not have the name expected, including the path.
   This affect to auto-load media (tapes, disks). In standard Fuse compilation cutoms roms don't autoload.
   The option "Auto-load media with custom rom" in General options (default False) auto-load also with custom roms.
 
-Also roms con be added to the 'rom' directory at Fuse config path ("$HOME/.fuse/roms").
+Also roms con be added to the 'rom' directory in config path ("$HOME/.fuse/roms").
 Fuse search in this path for rom files with the default names.
 
     For OpenDingux   "/media/data/local/home/.fuse/roms"
diff --git a/compat/linux/regex.c b/compat/linux/regex.c
index 0b96e0f..094dd47 100644
--- a/compat/linux/regex.c
+++ b/compat/linux/regex.c
@@ -47,7 +47,7 @@
 
   i = 0;
   while ( re_expressions[i] ) {
-    strncpy(new_search, cut_patterns( re_expressions[i], new_search ), MAX_LENGTH);
+    strncpy( new_search, cut_patterns( re_expressions[i], new_search ), MAX_LENGTH );
     i++;
   }
 
@@ -57,21 +57,21 @@
 static char*
 cut_patterns( const char* re_expression, const char* text )
 {
-  char remain[MAX_LENGTH ] = {'\0'};
+  char remain[MAX_LENGTH] = {'\0'};
   char *new_search;
   regex_t reg_expr;
 
-  new_search = libspectrum_new(char, MAX_LENGTH);
-  strncpy(new_search, text, MAX_LENGTH);
+  new_search = libspectrum_new( char, MAX_LENGTH );
+  strncpy( new_search, text, MAX_LENGTH-1 );
 
-  if (regcomp(&reg_expr, re_expression, REG_ICASE|REG_EXTENDED|REG_NEWLINE))
+  if ( regcomp( &reg_expr, re_expression, REG_ICASE|REG_EXTENDED|REG_NEWLINE ) )
     return new_search;
 
   while ( !cut_pattern( &reg_expr, new_search, &remain[0] ) )
     strncpy( new_search, &remain[0], 100 );
 
-  strncpy(new_search, &remain[0], 100);
-  regfree(&reg_expr);
+  strncpy( new_search, &remain[0], 100 );
+  regfree( &reg_expr );
 
   return new_search;
 }
@@ -83,14 +83,14 @@
 
   if ( !text ) return 1;
 
-  if ( regexec(reg_expr, text, MAX_MATCHES, matches, 0) ) {
-    strncpy( remain, text, MAX_LENGTH);
+  if ( regexec( reg_expr, text, MAX_MATCHES, matches, 0 ) ) {
+    strncpy( remain, text, MAX_LENGTH );
     return 1;
   }
 
   strncpy( remain, text, matches[0].rm_so );
   remain[matches[0].rm_so] = '\0';
-  if ( text[matches[0].rm_eo] ) strncat(remain, &text[matches[0].rm_eo], MAX_LENGTH);
+  if ( text[matches[0].rm_eo] ) strncat( remain, &text[matches[0].rm_eo], MAX_LENGTH-1 );
 
   return 0;
 }
\ No newline at end of file
diff --git a/controlmapping/controlmapping.c b/controlmapping/controlmapping.c
index c5c09de..b32507d 100644
--- a/controlmapping/controlmapping.c
+++ b/controlmapping/controlmapping.c
@@ -250,9 +250,10 @@
     controlmapping_save_current_mapfile();
 
     /* Not detached defaults: diabled. Save defaults and set current settings as new defaults */
-    if ( !settings_old.control_mapping_not_detached_defaults )
+    if ( !settings_old.control_mapping_not_detached_defaults ) {
       controlmapping_save_default_mapfile();
       controlmapping_reset_to_defaults( settings );
+    }
 
     /* Unassign mapfile */
     controlmapping_load_mapfile_with_class( NULL, LIBSPECTRUM_CLASS_UNKNOWN, 1 );
diff --git a/controlmapping/settings.pl b/controlmapping/settings.pl
index e162d0f..2734f50 100644
--- a/controlmapping/settings.pl
+++ b/controlmapping/settings.pl
@@ -118,7 +118,7 @@
 
 const char* re_expressions[] = {
     "(([[:space:]]|[-_])*)(([[]|[(])+[[:space:]]*)(([[:alnum:]]|[[:space:]]|[[:punct:]])*)([[:space:]]*([]]|[)])+)(([[:space:]]|[-_])*)",
-    "(([[:space:]]|[-_])*)(([(]|[[])*[[:space:]]*)(disk|tape|side|part)(([[:space:]]|[[:punct:]])*)(([abcd1234])([[:space:]]*of[[:space:]]*[1234])*)([[:space:]]*([)]|[]])*)(([[:space:]]|[-_])*)",
+    "(([[:space:]]|[-_])*)(([(]|[[])*[[:space:]]*)(disk|tape|side|part|release)(([[:space:]]|[[:punct:]])*)(([abcd1234])([[:space:]]*of[[:space:]]*[1234])*)([[:space:]]*([)]|[]])*)(([[:space:]]|[-_])*)",
     "(([[:space:]]|[-_])*)((128k|48k)+)(([[:space:]]|[-_])*)",
     "(([[:space:]]|[-_])*)((small|medium|large)[[:space:]]*[[:alnum:]]*[[:space:]]*case)(([[:space:]]|[-_])*)",
     NULL };
@@ -769,7 +769,10 @@
   /* Don't exist config path, no error but do nothing */
   cfgdir = compat_get_config_path(); if( !cfgdir ) return NULL;
 
-  filename_test = compat_chop_expressions( re_expressions, utils_last_filename( filename, 1 ) );
+  if ( settings_current.control_mapping_no_transform_filename )
+    filename_test = utils_last_filename( filename, 1 );
+  else
+    filename_test = compat_chop_expressions( re_expressions, utils_last_filename( filename, 1 ) );
 
   snprintf( buffer, PATH_MAX, "%s"FUSE_DIR_SEP_STR"%s"FUSE_DIR_SEP_STR"%s%s",
             cfgdir, "mappings", filename_test, ".fcm" );
diff --git a/fuse.c b/fuse.c
index 7853e9a..4e8f035 100644
--- a/fuse.c
+++ b/fuse.c
@@ -389,10 +389,6 @@
 
   if( settings_init( &first_arg, argc, argv ) ) return 1;
 
-#ifdef GCWZERO
-  settings_current.full_screen = 1;
-#endif
-
   if( settings_current.show_version ) {
     fuse_show_version();
     return 0;
diff --git a/menu.c b/menu.c
index 5775816..e441432 100644
--- a/menu.c
+++ b/menu.c
@@ -1147,10 +1147,7 @@
 const char*
 menu_control_mapping_detail( void )
 {
-  if ( settings_current.control_mapping_per_game )
-    return "Active";
-  else
-    return "Inactive";
+  return settings_current.control_mapping_per_game ? "Enabled" : "Disabled";
 }
 
 const char*
diff --git a/menu.h b/menu.h
index 7b8f441..ed428e1 100644
--- a/menu.h
+++ b/menu.h
@@ -164,6 +164,7 @@
 MENU_DETAIL( menu_joystick_1_detail );
 MENU_DETAIL( menu_joystick_2_detail );
 #ifdef GCWZERO
+MENU_CALLBACK( menu_options_generalgcw0 );
 MENU_DETAIL( menu_gcw0_keyboard_detail );
 MENU_DETAIL( menu_control_mapping_detail );
 MENU_DETAIL( menu_control_mapping_load_detail );
diff --git a/menu_data.dat b/menu_data.dat
index c55855b..0d24592 100644
--- a/menu_data.dat
+++ b/menu_data.dat
@@ -76,6 +76,7 @@
 
 _Options, Branch
 Options/_General..., Item, F4
+Options/General GC_W0..., Item
 Options/_Media..., Item
 Options/_Sound..., Item
 
diff --git a/peripherals/joystick.c b/peripherals/joystick.c
index cfaf463..24707ff 100644
--- a/peripherals/joystick.c
+++ b/peripherals/joystick.c
@@ -79,7 +79,7 @@
 
 #ifdef GCWZERO
 const char *joystick_name_gcw0[] = {
-  "None", "Activated"
+  "None", "Enabled"
 };
 #endif
 
diff --git a/settings.dat b/settings.dat
index cbad031..be3630b 100644
--- a/settings.dat
+++ b/settings.dat
@@ -60,6 +60,7 @@
 confirm_overwrite_files, boolean, 0
 hidden_files, boolean, 0
 hotkey_combos, boolean, 0
+show_fps, boolean, 0
 printer, boolean, 0
 statusbar, boolean, 1
 interface1, boolean, 0
@@ -171,6 +172,7 @@
 control_mapping_autoload, boolean, 1
 control_mapping_autosave, boolean, 1
 control_mapping_not_detached_defaults, boolean, 0
+control_mapping_no_transform_filename, boolean, 0
 
 rzx_compression, boolean, 1,, compress-rzx
 competition_mode, boolean, 0
diff --git a/timer/timer.c b/timer/timer.c
index f99536f..96b7ec5 100644
--- a/timer/timer.c
+++ b/timer/timer.c
@@ -97,11 +97,13 @@
   }
 
 #ifdef GCWZERO
-  ui_statusbar_update_speed( frames_per_second );
-  frame_count = 0;
-#else
-  ui_statusbar_update_speed( current_speed );
+  if ( settings_current.show_fps ) {
+    ui_statusbar_update_speed( frames_per_second );
+    frame_count = 0;
+  } else
 #endif
+  ui_statusbar_update_speed( current_speed );
+
   stored_times[ next_stored_time ] = current_time;
 
   next_stored_time = ( next_stored_time + 1 ) % 10;
diff --git a/ui.c b/ui.c
index c1b949d..17ae22f 100644
--- a/ui.c
+++ b/ui.c
@@ -757,7 +757,7 @@
   /* Add the extension .TZX */
   filename = realloc( filename, strlen( filename ) + 4 );
   if( !filename ) { fuse_emulation_unpause(); return 1; }
-  strcat( filename, ".tzx" );
+  strncat( filename, ".tzx", 4 );
 #endif
 
   tape_write( filename );
@@ -787,7 +787,7 @@
     /* Add the extension .MDR */
     filename = realloc( filename, strlen( filename ) + 4 );
     if( !filename ) { fuse_emulation_unpause(); return 1; }
-    strcat( filename, ".mdr" );
+    strncat( filename, ".mdr", 4 );
 #endif
   }
 
diff --git a/ui/options.dat b/ui/options.dat
index ee17064..9c255c1 100644
--- a/ui/options.dat
+++ b/ui/options.dat
@@ -18,19 +18,21 @@
 #ifndef GCWZERO
 Checkbox, Full (s)creen, full_screen, INPUT_KEY_s
 #endif
-#ifdef GCWZERO
-Checkbox, Triple Bu(f)fer, triple_buffer, INPUT_KEY_f
-#endif
 #endif
 Checkbox, Show status(b)ar, statusbar, INPUT_KEY_b
 Checkbox, Snap (j)oystick prompt, joy_prompt, INPUT_KEY_j
 Checkbox, (C)onfirm actions, confirm_actions, INPUT_KEY_c
+Checkbox, A(u)to-save settings, autosave_settings, INPUT_KEY_u
+
 #ifdef GCWZERO
+general_gcw0
+General GCW0 Options
+Checkbox, Triple Bu(f)fer, triple_buffer, INPUT_KEY_f
+Checkbox, Sho(w) FPS instead of speed percentaje, show_fps, INPUT_KEY_w
 Checkbox, Confir(m) overwrite files, confirm_overwrite_files, INPUT_KEY_m
 Checkbox, Show hi(d)den files, hidden_files, INPUT_KEY_d
 Checkbox, Hot(k)ey combos, hotkey_combos, INPUT_KEY_k
 #endif
-Checkbox, A(u)to-save settings, autosave_settings, INPUT_KEY_u
 
 media
 Media Options
@@ -153,4 +155,5 @@
 Checkbox, Auto-(l)oad, control_mapping_autoload, INPUT_KEY_L
 Checkbox, Auto-(s)ave, control_mapping_autosave, INPUT_KEY_S
 Checkbox, No(t) detached defaults, control_mapping_not_detached_defaults, INPUT_KEY_T
+Checkbox, No cut/transfor(m) filenames, control_mapping_no_transform_filename, INPUT_KEY_M
 #endif
\ No newline at end of file
diff --git a/ui/sdl/sdldisplay.c b/ui/sdl/sdldisplay.c
index 77227d2..8397d50 100644
--- a/ui/sdl/sdldisplay.c
+++ b/ui/sdl/sdldisplay.c
@@ -261,6 +261,88 @@
   return 0;
 }
 
+#ifdef GCWZERO
+/* Initializations for OpenDingux/RetroFW */
+void
+uidisplay_od_init( SDL_Rect **modes )
+{
+  char line[100];
+  char* ptok;
+
+/* On OpenDingux/RetroFW fix Full Screen */
+  settings_current.full_screen = 1;
+  sdldisplay_change_triple_buffer = settings_current.triple_buffer;
+  settings_current.sdl_fullscreen_mode = utils_safe_strdup( '\0' );
+
+/*
+ * Dirty hack. There should be a better solution.
+ * In RetroFW 1 the video driver don't downscale correctly. This is problematic
+ * with filters higher than 1x, the can be choosed but the image is not correct.
+ * This should prevent to use any filter that will be problematic.
+ * Also for Timex Machines this will force to use the Timex Half filters...
+ *
+ * First:
+ * for OpenDingux devices like GCW0, RG350 the info about downscaling is at:
+ *        /sys/devices/platform/jz-lcd.0/allow_downscaling
+ *        If it can be read then it's done.
+ *
+ * Second:
+ * Read /etc/os-release. RetroFW 2 downscale right, so if NAME is Buildroot then
+ * the full_screen_mode will be fixed to 320x240.
+ *
+ * This avoids that the use of scalers with a scale factor greater than 1x
+ * which in RetroFW1 makes it try to use the actual resolution and not downscaling
+ * to the resolution of display.
+ *
+ * But probably this will be problematic on devices with no 320x240 resolution.
+ *
+ *  For RetroFW 2.x
+ *    NAME=RetroFW
+ *    ID=retrofw
+ *    VERSION_ID=2.2
+ *    PRETTY_NAME="RetroFW v2.2"
+ *    ANSI_COLOR="0;36"
+ *    HOME_URL="https://retrofw.github.io/"
+ *    BUG_REPORT_URL="https://github.com/retrofw/retrofw.github.io/issues"
+ *
+ * For RetroFW 1
+ *    NAME=Buildroot
+ *    VERSION=2018.02.11
+ *    ID=buildroot
+ *    VERSION_ID=2018.02.11
+ *    PRETTY_NAME="Buildroot 2018.02.11"
+*
+ * For OpenDingux:
+ *    NAME=Buildroot
+ *    VERSION=2014.08-g156cb719e
+ *    ID=buildroot
+ *    VERSION_ID=2014.08
+ *    PRETTY_NAME="Buildroot 2014.08"
+ */
+  FILE* allow_downscaling = fopen( "/sys/devices/platform/jz-lcd.0/allow_downscaling", "r" );
+  if ( !allow_downscaling ) {
+    /* We are on RetroFW */
+    FILE* os_release = fopen( "/etc/os-release", "r" );
+    if ( os_release ) {
+      while ( fgets(line, sizeof( line ), os_release ) != NULL ) {
+        ptok = strtok( line, "=" );
+        if ( strcmp( ptok, "NAME" ) == 0 ) {
+          ptok = strtok( NULL, "=" );
+          /* And we are on RetroFW 1.X */
+          if ( strcmp( ptok, "Buildroot\n" ) == 0)
+            settings_current.sdl_fullscreen_mode = utils_safe_strdup( "320x240" );
+          break;
+        }
+      }
+      fclose( os_release );
+    }
+
+  /* We are on OpenDingux */
+  } else
+    fclose( allow_downscaling );
+}
+#endif
+
 int
 uidisplay_init( int width, int height )
 {
@@ -278,34 +360,7 @@
   no_modes = ( modes == (SDL_Rect **) 0 || modes == (SDL_Rect **) -1 ) ? 1 : 0;
 
 #ifdef GCWZERO
-  settings_current.full_screen = 1;
-  sdldisplay_change_triple_buffer = settings_current.triple_buffer;
-  settings_current.sdl_fullscreen_mode = utils_safe_strdup( '\0' );
-/*
- * Dirty hack. There should be a better solution.
- * In RetroFW 1 the video driver don't downscale correctly. This is problematic
- * with filters higher than 1x, the can be choosed but the image is not correct.
- * This should prevent to use any filter that will be problematic.
- * Also for Timex Machines this will force to use the Timex Half filters...
- */
-  FILE* allow_downscaling = fopen("/sys/devices/platform/jz-lcd.0/allow_downscaling","r");
-  if (!allow_downscaling) {
-    FILE* os_release = fopen("/etc/os-release", "r");
-    if (os_release) {
-      char line[100];
-      char* ptok;
-      while ( fgets(line,sizeof(line),os_release) != NULL ) {
-        ptok = strtok(line,"=");
-        if (strcmp(ptok,"NAME") == 0) {
-          ptok = strtok(NULL,"=");
-          if (strcmp(ptok,"Buildroot\n") == 0)
-            settings_current.sdl_fullscreen_mode = utils_safe_strdup( "320x240" );
-          break;
-        }
-      }
-      fclose(os_release);
-    }
-  } else fclose(allow_downscaling);
+  uidisplay_od_init( modes );
 #endif
 
   if( settings_current.sdl_fullscreen_mode &&
diff --git a/ui/widget/menu.c b/ui/widget/menu.c
index 71daf9f..a32d99a 100644
--- a/ui/widget/menu.c
+++ b/ui/widget/menu.c
@@ -784,7 +784,7 @@
   submenu_types[ i + 1 ].text = NULL;
 
 #ifdef GCWZERO
-  submenu_types_gcw0[ 0 ].text = "Activate joystick";
+  submenu_types_gcw0[ 0 ].text = "Enable joystick";
   for( i = 0; i < 2; i++ ) {
     char shortcut[ 2 ] = { 'A' + i, '\0' };
     snprintf( ( char * ) joystick_names_gcw0[ i ], 100, "\012%s\011 %s", shortcut,
@@ -827,6 +827,12 @@
 
 #ifdef GCWZERO
 void
+menu_options_generalgcw0( int action )
+{
+  widget_do_generalgcw0();
+}
+
+void
 menu_options_joysticks_controlmappingpergame( int action )
 {
   widget_do_joysticks_controlmapping();
diff --git a/ui/widget/widget.c b/ui/widget/widget.c
index 081ecf6..28e3074 100644
--- a/ui/widget/widget.c
+++ b/ui/widget/widget.c
@@ -546,7 +546,7 @@
 }
 
 #ifdef GCWZERO
-#define WIDGET_MAX_INFO_LENGTH 40
+#define WIDGET_MAX_INFO_LENGTH 60
 #define WIDGET_COL_INFO 5
 #define WIDGET_ROW_INFO 225
 #define WIDGET_RELATIVE_COL_INFO (WIDGET_COL_INFO - DISPLAY_BORDER_ASPECT_WIDTH)
@@ -555,16 +555,20 @@
 static int  info_len = 0;
 static int  info_sc  = 1;
 void widget_statusbar_update_info( float speed ) {
-/*
-  snprintf(status_info, WIDGET_MAX_INFO_LENGTH, "%s - %3.0f%% (1:%d)",
+  snprintf(status_info, WIDGET_MAX_INFO_LENGTH,
+           settings_current.show_fps ? "%s - %3.0ffps (1:%d)" : "%s - %3.0f%% (1:%d)",
            libspectrum_machine_name( machine_current->machine ),
            speed,
            settings_current.frame_rate);
-*/
-  snprintf(status_info, WIDGET_MAX_INFO_LENGTH, "%s - %3.0ffps (1:%d)",
-           libspectrum_machine_name( machine_current->machine ),
-           speed,
-           settings_current.frame_rate);
+  if ( settings_current.joystick_1_output || settings_current.joystick_gcw0_output )
+    snprintf(status_info, WIDGET_MAX_INFO_LENGTH, "%s [%s]",
+             status_info,
+             settings_current.joystick_1_output ? joystick_name[settings_current.joystick_1_output]
+                                                : "Keyboard" );
+  if ( settings_current.joystick_2_output )
+    snprintf(status_info, WIDGET_MAX_INFO_LENGTH, "%s [%s]",
+             status_info,
+             joystick_name[settings_current.joystick_2_output] );
   if ( settings_current.triple_buffer )
     snprintf(status_info, WIDGET_MAX_INFO_LENGTH, "%s [%s]",
              status_info,
@@ -813,6 +817,7 @@
 #endif
 #ifdef GCWZERO
   { widget_control_mapping_draw, widget_control_mapping_finish, widget_control_mapping_keyhandler  },
+  { widget_general_gcw0_draw,  widget_options_finish, widget_general_gcw0_keyhandler  },
 #endif
 };
 
diff --git a/ui/widget/widget.h b/ui/widget/widget.h
index 5a1364e..1f1f1cc 100644
--- a/ui/widget/widget.h
+++ b/ui/widget/widget.h
@@ -81,6 +81,7 @@
 #endif
 #ifdef GCWZERO
   WIDGET_TYPE_JOYSTICKS_CONTROLMAPPING, /* Control mapping */
+  WIDGET_TYPE_GENERALGCW0,		/* General GCW0 options */
 #endif
 } widget_type;
 
@@ -184,6 +185,12 @@
 #endif
 
 #ifdef GCWZERO
+/* General options for GCW0 */
+static inline int widget_do_generalgcw0( void )
+{
+  return widget_do( WIDGET_TYPE_GENERALGCW0, NULL );
+}
+
 /* Control Mapping */
 static inline int widget_do_joysticks_controlmapping( void )
 {
